src/core/geometry/qgsgeometry.cpp

```diff
@@ -1449,60 +1450,176 @@ QVector<QgsGeometry> QgsGeometry::coerceToType( const Qgis::WkbType type, double
     newGeom = QgsGeometry( d->geometry.get()->segmentize() );
   }
 
+  // compoundcurve -> circularstring (possible if compoundcurve consists of one circular string)
+  if ( QgsWkbTypes::flatType( type ) == Qgis::WkbType::CircularString )
```

Comment:
Using QgsAbstractGeometry::simplifiedTypeRef() might help a lot here, a lot of this logic is already in place there.